### PR TITLE
FIx tf keras inferenc issue for rest server

### PR DESCRIPTION
**Pull request by yubozhao**
What changes were proposed in this pull request?
------------------------------------------------
Fixed issue https://github.com/bentoml/BentoML/issues/94.

*Issue:*  
Tensorflow Keras model not able to predict in Flask app. With Error: `Tensor Tensor("dense_2/Sigmoid:0", shape=(?, 1), dtype=float32) is not an element of this graph.`

*Possible Cause:*
The Tensorflow keras model loading and prediction might not be in the same thread, because flask's usage of multi thread.  

*Solution:*
Create a TF model wrapper that will load and predict with the same session and graph.

References:
* https://stackoverflow.com/questions/51127344/tensor-is-not-an-element-of-this-graph-deploying-keras-model
* https://github.com/tensorflow/tensorflow/issues/14356

Does this close any currently open issues?
------------------------------------------

How was this patch tested?
--------------------------
https://github.com/bentoml/BentoML/issues/94

### CODE REVIEW

File: bentoml/artifact/tf_keras_model_artifact.py [hunk 1]
```diff
@@ -68,3 +69,30 @@ def save(self, dst):
...
+            with self.session.as_default():
+                self.model = load_model(model)
+
+    def predict(self, x):
+        with self.graph.as_default():
+            with self.session.as_default():
+                result = self.model.predict(x)
+        return result
+
+    def predict_classes(self, x):
```
Comment (parano) [inline bentoml/artifact/tf_keras_model_artifact.py]: this is not a `training.Model` method, only `Sequential` has it

Comment (parano) [reply bentoml/artifact/tf_keras_model_artifact.py]: also make sure to pass down all parameters for this type of wrapper  `def predict_classes(self, x, batch_size=32, verbose=0):`

File: bentoml/artifact/tf_keras_model_artifact.py [hunk 2]
```diff
@@ -68,3 +69,30 @@ def save(self, dst):
...
+    def __init__(self, model):
+        try:
+            import tensorflow as tf
+            from tensorflow.python.keras.models import load_model
+        except ImportError:
+            raise ImportError("tensorflow package is required to use TfKerasModelArtifact")
+        self.session = tf.Session()
+        self.graph = tf.get_default_graph()
+        with self.graph.as_default():
+            with self.session.as_default():
```
Comment (parano) [inline bentoml/artifact/tf_keras_model_artifact.py]: is this necessary?

File: bentoml/artifact/tf_keras_model_artifact.py (hunk 1 above)
Comment (parano) [reply bentoml/artifact/tf_keras_model_artifact.py]: I think might be better to force gloabl default graph for entire API function instead of wrapping individual keras model methods here

Comment (yubozhao) [reply bentoml/artifact/tf_keras_model_artifact.py]: I was afraid of using it as global default graph, but I will test that route.

File: bentoml/artifact/tf_keras_model_artifact.py (hunk 2 above)
Comment (yubozhao) [reply bentoml/artifact/tf_keras_model_artifact.py]: I think so, you need to load the model in the same session/graph

File: bentoml/artifact/tf_keras_model_artifact.py (hunk 1 above)
Comment (yubozhao) [reply bentoml/artifact/tf_keras_model_artifact.py]: Another pro of wrapping the individual keras model is there is no additional actions from user.  If we use global graph, user have to make prediction call within that graph context.  I think it would be just strange to users to use a var that 'comes out' from nowhere.

```python
...
def predict(self, df):
      with graph.as_default():    # user has no knowledge of where this come from.
              result = self.artifacts.model.predict(df)
```

File: bentoml/artifact/tf_keras_model_artifact.py [hunk 3]
```diff
@@ -68,3 +69,30 @@ def save(self, dst):
...
+
+class _TfKerasModelWrapper(object):
+
+    def __init__(self, model):
+        try:
+            import tensorflow as tf
+            from tensorflow.python.keras.models import load_model
+        except ImportError:
+            raise ImportError("tensorflow package is required to use TfKerasModelArtifact")
+        self.session = tf.Session()
```
Comment (parano) [inline bentoml/artifact/tf_keras_model_artifact.py]: shouldn't this be:
```
self.graph = tf.get_default_graph()
sess = tf.Session(graph=graph)
```?

File: bentoml/artifact/tf_keras_model_artifact.py (hunk 1 above)
Comment (parano) [reply bentoml/artifact/tf_keras_model_artifact.py]: @yubozhao oh I didn't mean "global" graph, I was referring to self.graph here. But instead of wrapping on Keras model, wrap it around the API function.

Comment (parano) [reply bentoml/artifact/tf_keras_model_artifact.py]: Btw, I have some new changes on how user API function access artifacts, that might make this implementation easier.

Comment (yubozhao) [reply bentoml/artifact/tf_keras_model_artifact.py]: Oh. I got you.  Also. I totally forgot to mention to you that this PR is not mean to be merged. I was able to get the tf keras model working with this, but I don’t think this is a good solution and there are going to be potential problems down the line. This PR’s goal is start to have a conversation around this
